source/fuzz/transformation_changing_memory_semantics.cpp

```diff
@@ -0,0 +1,296 @@
+  // of instruction that takes one memory semantics operand and 0 or 1 for
+  // instructions that takes two operands.
+  if (!IsNeededOpcodeWithAppropriateIndex(
+          needed_atomic_instruction->opcode(),
+          message_.memory_semantics_operand_index())) {
+    return false;
+  }
+
+  // The new value instruction must exist and must be OpConstant.
+  auto value_instruction = ir_context->get_def_use_mgr()->GetDef(
+      message_.memory_semantics_new_value_id());
+  if (!value_instruction) {
+    return false;
+  }
+  if (value_instruction->opcode() != SpvOpConstant) {
+    return false;
+  }
+
+  // The first 5 bits of new memory semantics values must be suitable for needed
+  // instruction. The first 5 bits of new value must be larger than the first 5
+  // bits of old. Can't use Sequentially Consistent memory semantic if the
+  // memory model is Vulkan.
+  auto new_memory_sematics_value = value_instruction->GetSingleWordInOperand(0);
+  uint32_t old_memory_sematics_value = 0;
+  if (needed_atomic_instruction->opcode() == SpvOpMemoryBarrier) {
+    // Memory semantics true index for the OpMemoryBarrier equal 1.
+    old_memory_sematics_value =
+        ir_context->get_def_use_mgr()
+            ->GetDef(needed_atomic_instruction->GetSingleWordInOperand(1))
+            ->GetSingleWordInOperand(0);
+  } else {
+    old_memory_sematics_value =
+        ir_context->get_def_use_mgr()
+            ->GetDef(needed_atomic_instruction->GetSingleWordInOperand(
+                message_.memory_semantics_operand_index() == 0 ? 2 : 3))
+            ->GetSingleWordInOperand(0);
+  }
+  auto first_5bits_new_memory_semantics =
+      static_cast<SpvMemorySemanticsMask>(new_memory_sematics_value & 0x1F);
+  auto first_5bits_old_memory_semantics =
+      static_cast<SpvMemorySemanticsMask>(old_memory_sematics_value & 0x1F);
+  auto memory_model = static_cast<SpvMemoryModel>(
+      ir_context->module()->GetMemoryModel()->GetSingleWordInOperand(1));
+
+  if (!IsValidConverstion(needed_atomic_instruction->opcode(),
+                          first_5bits_old_memory_semantics,
+                          first_5bits_new_memory_semantics, memory_model)) {
+    return false;
+  }
+
+  // The higher bits value of old and new memory semantics id must be equal.
+  auto higher_bits_new_memory_semantics = static_cast<SpvMemorySemanticsMask>(
+      new_memory_sematics_value & 0xFFFFFFE0);
+  auto higher_bits_old_memory_semantics = static_cast<SpvMemorySemanticsMask>(
+      old_memory_sematics_value & 0xFFFFFFE0);
+  if (higher_bits_new_memory_semantics != higher_bits_old_memory_semantics) {
+    return false;
+  }
+
+  // Instructions that takes two memory semantics, one of the value must be
+  // stronger than the other.
+  if (needed_atomic_instruction->opcode() == SpvOpAtomicCompareExchange ||
+      needed_atomic_instruction->opcode() == SpvOpAtomicCompareExchangeWeak) {
+    auto other_memory_semantics_index = static_cast<SpvMemorySemanticsMask>(
```

Comment:
This variable name is confusing. I don't think it is the operand index. It is the memory semantics value.

Reply:
Changed.